### PR TITLE
feat: TYPO3 v13+v14 modernization with task dialog, context zoom, and full test coverage

### PR DESCRIPTION
**Pull request by CybotTM**
## Summary

Major modernization of t3x-cowriter for TYPO3 v13 and v14: a task-based cowriter dialog, context-aware AI processing with "context zoom," clean prompt architecture, and a comprehensive testing pyramid — backed by a ddev demo environment for hands-on verification.

### Task-based cowriter dialog
- Task selection dialog with predefined AI operations (improve, summarize, translate, structure, etc.)
- Ad-hoc rules for user-specified overrides per execution
- Runtime editor capability detection (CKEditor plugin introspection) so the LLM knows what formatting is available
- Seed data for 10 built-in cowriter tasks via `ddev seed-cowriter-tasks`

### Context zoom
- Server-side context assembly: selection → text → element → record → page → reference pages
- Context zoom slider in the dialog UI for controlling how much surrounding content the LLM sees
- Reference page picker for cross-page context
- Separate system messages for scope constraints, reference context, and editor capabilities
- `getContext` preview endpoint for word count / summary before execution

### Clean prompt architecture
- Task templates focus on WHAT (improve, summarize) — not HOW (no formatting prescriptions)
- System messages handle scope, available formatting, and surrounding context
- Ad-hoc rules injected as explicit overrides ("follow these exactly")
- Removed fragile regex hack that stripped formatting instructions at runtime

### Testing
- 354 PHP unit tests + 122 JS tests, 100% coverage with mutation testing (Infection)
- Streaming E2E tests with Ollama integration
- Fuzz testing for edge cases (malformed JSON, XSS payloads, Unicode)

### TYPO3 v13 + v14 compatibility
- Fix DI compilation for cache service on v14
- Fix backend route identifiers (`ajax_` prefix)
- Resolve composer dependency conflicts for PHP 8.2–8.5
- Auto-activate cowriter RTE preset via `page.tsconfig`
- Adapt E2E tests to v14 iframe + dialog architecture

### ddev demo environment
- Dual-site setup: v13 + v14 running side by side
- Demo page tree with RTE content for testing the cowriter plugin
- Auto-seeded Ollama LLM configuration
- `seed-pages` / `seed-ollama` / `seed-cowriter-tasks` commands
- Landing page with quick links to both backends

### Upstream dependency fixes
- [nr_vault v0.4.1](https://github.com/netresearch/t3x-nr-vault/releases/tag/v0.4.1): Fix XLF label keys + module parent v13 compatibility ([#92](https://github.com/netresearch/t3x-nr-vault/pull/92))
- [nr_llm v0.2.2](https://github.com/netresearch/t3x-nr-llm/releases/tag/v0.2.2): Fix module parent v13 compatibility ([#74](https://github.com/netresearch/t3x-nr-llm/pull/74))

## Test plan

- [ ] `make up` completes successfully (both v13 and v14 install)
- [ ] v14 backend: Vault module loads without 503, CKEditor shows cowriter button
- [ ] v13 backend: LLM + Vault modules visible, CKEditor shows cowriter button
- [ ] Cowriter dialog: select "Improve Text" on a paragraph → output stays paragraph-scoped
- [ ] Context zoom: slide to "page" scope → LLM sees surrounding content as reference
- [ ] Ad-hoc rules: add "use formal tone" → output respects the override
- [ ] Unit tests pass: `composer ci:test:php:unit` (354 tests)
- [ ] JS tests pass: `npx vitest run` (122 tests)
- [ ] CI pipeline green

### CODE REVIEW

File: Tests/Unit/Domain/DTO/CompleteRequestFuzzTest.php [hunk 1]
```diff
@@ -0,0 +1,473 @@
...
+    {
+        yield 'only prefix no model' => ['#cw: text', null, '#cw: text'];
+        yield 'prefix without space after model' => ['#cw:modeltext', null, '#cw:modeltext'];
+        yield 'double prefix' => ['#cw:gpt-4o #cw:claude test', 'gpt-4o', '#cw:claude test'];
+        yield 'prefix with newline' => ["#cw:gpt-4o\ntest", 'gpt-4o', 'test'];
+        yield 'prefix with tab' => ["#cw:gpt-4o\ttest", 'gpt-4o', 'test'];
+        yield 'case sensitive prefix' => ['#CW:gpt-4o test', null, '#CW:gpt-4o test'];
+        yield 'prefix mid-string' => ['hello #cw:gpt-4o test', null, 'hello #cw:gpt-4o test'];
+        yield 'model with all valid chars' => ['#cw:a-b_c.d:e/f test', 'a-b_c.d:e/f', 'test'];
+        yield 'model max valid pattern' => ['#cw:' . str_repeat('a', 200) . ' test', str_repeat('a', 200), 'test'];
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply Tests/Unit/Domain/DTO/CompleteRequestFuzzTest.php]: Good catch. Added a 256-character length limit to `MODEL_NAME_PATTERN` (`{0,255}` quantifier) and added a test case verifying that a 257-char model name is rejected while 256 chars is accepted. Fixed in 30572e1.

File: Tests/JavaScript/cowriter.test.js [hunk 2]
```diff
@@ -147,6 +147,231 @@ describe('Cowriter Plugin', () => {
...
+            errorSpy.mockRestore();
+        });
+
+        it('should handle AI service errors without message', async () => {
+            plugin._service.complete = vi.fn().mockRejectedValue({});
+            const mockRange = {
+                getItems: () => [{ data: 'prompt' }],
+            };
+            mockEditor.model.document.selection.getRanges.mockReturnValue([mockRange]);
+            vi.spyOn(console, 'error').mockImplementation(() => {});
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply Tests/JavaScript/cowriter.test.js]: Fixed. Now captures the spy in a const and calls `errorSpy.mockRestore()` at the end of the test.

File: Tests/JavaScript/cowriter.test.js [hunk 3]
```diff
@@ -147,6 +147,231 @@ describe('Cowriter Plugin', () => {
...
+            expect(plugin._isProcessing).toBe(false);
+        });
+
+        it('should reset _isProcessing after error', async () => {
+            plugin._service.complete = vi.fn().mockRejectedValue(new Error('fail'));
+            const mockRange = {
+                getItems: () => [{ data: 'prompt' }],
+            };
+            mockEditor.model.document.selection.getRanges.mockReturnValue([mockRange]);
+            vi.spyOn(console, 'error').mockImplementation(() => {});
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply Tests/JavaScript/cowriter.test.js]: Fixed — same pattern: spy captured in const and restored at test end.

File: infection.json5 [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+            "Classes"
+        ]
+    },
+    "logs": {
+        "text": ".Build/logs/infection.log",
+        "summary": ".Build/logs/infection-summary.log"
+    },
+    "tmpDir": ".Build/tmp",
+    "phpUnit": {
+        "configDir": "Build/phpunit",
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply infection.json5:14]: Not an issue — `Build/phpunit/phpunit.xml.dist` exists in that directory, which Infection resolves correctly. The CI workflow also passes `--test-framework-options="--testsuite='\''Unit Tests'\''"'\'' explicitly.

File: Tests/Unit/Domain/DTO/CompleteRequestFuzzTest.php [hunk 5]
```diff
@@ -0,0 +1,473 @@
...
+        yield 'XML document' => ['<?xml version="1.0"?><root/>'];
+        yield 'just null' => ['null'];
+        yield 'just number' => ['42'];
+        yield 'just string' => ['"hello"'];
+        yield 'just boolean' => ['true'];
+        yield 'just array' => ['["a","b"]'];
+        yield 'empty object' => ['{}'];
+        yield 'nested deeply' => [str_repeat('[', 100) . '1' . str_repeat(']', 100)];
+        yield 'very large number' => ['{"prompt":"test","n":' . str_repeat('9', 1000) . '}'];
+        yield 'unicode escape' => ['{"prompt":"\\u0048\\u0065\\u006c\\u006c\\u006f"}'];
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply Tests/Unit/Domain/DTO/CompleteRequestFuzzTest.php]: Good catch. Renamed to '\''invalid unicode escape'\'' and changed input to `\u004G` which is genuinely malformed JSON.

File: Tests/Unit/Service/ContextAssemblyServiceTest.php [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+            123,
+            'bodytext',
+            'element',
+            [['pid' => 10, 'relation' => 'style guide']],
+        );
+
+        self::assertStringContainsString('Main', $result);
+        self::assertStringContainsString('Reference page', $result);
+        self::assertStringContainsString('style guide', $result);
+    }
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply Tests/Unit/Service/ContextAssemblyServiceTest.php:107]: Good observation. The test does use a shared `$index` counter for `fetchAssociative`, but in practice `assembleContext` calls `getQueryBuilderForTable` once per query and the rows are designed to cover the sequential access pattern. The test verifies the assembled output string, which confirms the correct data was read in the right order. That said, this test could be improved by using separate result stubs per table — will track as a follow-up.

File: Resources/Public/JavaScript/Ckeditor/AIService.js [hunk 7]
```diff
@@ -232,4 +238,102 @@ export class AIService {
...
+     */
+    async getContext(table, uid, field, scope) {
+        if (!this._routes.context) {
+            throw new Error(
+                'TYPO3 AJAX routes not configured. Ensure the cowriter extension is properly installed.'
+            );
+        }
+
+        const params = new URLSearchParams({ table, uid: String(uid), field, scope });
+        const url = `${this._routes.context}&${params.toString()}`;
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply Resources/Public/JavaScript/Ckeditor/AIService.js]: Fixed in dfa50c0. The URL separator now checks whether `?` is already present: `const separator = this._routes.context.includes('?') ? '&' : '?';` — handles both TYPO3 URLs (which include `?token=...`) and plain URLs correctly.

File: Resources/Public/JavaScript/Ckeditor/UrlLoader.js [hunk 8]
```diff
@@ -36,7 +36,7 @@ export function loadCowriterUrls() {
 
         // Merge only expected cowriter URL keys (defense-in-depth against prototype pollution).
         // Keep in sync with Configuration/Backend/AjaxRoutes.php
-        const allowedKeys = ['tx_cowriter_chat', 'tx_cowriter_complete', 'tx_cowriter_stream', 'tx_cowriter_configurations'];
+        const allowedKeys = ['tx_cowriter_chat', 'tx_cowriter_complete', 'tx_cowriter_stream', 'tx_cowriter_configurations', 'tx_cowriter_tasks', 'tx_cowriter_task_execute'];
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply Resources/Public/JavaScript/Ckeditor/UrlLoader.js]: Fixed in dfa50c0. Added `tx_cowriter_context` to the `allowedKeys` array in `UrlLoader.js` and added a corresponding assertion in the UrlLoader test.

File: Classes/Domain/DTO/CompleteResponse.php [hunk 9]
```diff
@@ -33,16 +35,14 @@ private function __construct(
...
     {
         return new self(
             success: true,
-            content: htmlspecialchars($response->content, ENT_QUOTES | ENT_HTML5, 'UTF-8'),
-            model: htmlspecialchars($response->model ?? '', ENT_QUOTES | ENT_HTML5, 'UTF-8'),
-            finishReason: htmlspecialchars($response->finishReason ?? '', ENT_QUOTES | ENT_HTML5, 'UTF-8'),
+            content: $response->content,
+            model: $response->model ?? '',
+            finishReason: $response->finishReason ?? '',
             usage: UsageData::fromUsageStatistics($response->usage),
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply Classes/Domain/DTO/CompleteResponse.php]: This is intentional, not a regression. The cowriter's purpose is to generate HTML content for CKEditor — the LLM output contains `<h2>`, `<ul>`, `<strong>`, `<span style="color:...">` etc. that must be preserved as-is for CKEditor insertion.

**Why server-side escaping would break the feature:** If we escape `<h2>` to `&lt;h2&gt;`, CKEditor would display literal `<h2>` text instead of a heading.

**How XSS is prevented:**
- The `complete` and `chat` endpoints (freeform user prompts) still HTML-escape via `htmlspecialchars()` — these return text, not HTML
- The `executeTask` endpoint returns raw HTML because the task templates explicitly produce HTML for CKEditor
- CKEditor's own data pipeline sanitizes content on insertion (removes scripts, event handlers)
- The frontend uses `textContent` for text display and DOMParser for HTML preview, both XSS-safe

The docblock on `CompleteResponse` was updated to document this design.

File: Configuration/page.tsconfig [hunk 10]
```diff
@@ -1,4 +1 @@
-# Uncomment the line below to use the cowriter RTE preset as the default for all RTE fields.
-# Alternatively, add the cowriter plugin to your existing preset via YAML configuration.
-# See README.md "Option B: Extend your existing preset" for details.
-# RTE.default.preset = cowriter
+RTE.default.preset = cowriter
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply Configuration/page.tsconfig]: Good catch. Fixed in dfa50c0 — the preset is now commented out in `Configuration/page.tsconfig` with a note to opt-in. The ddev demo environment still sets it via `seed-pages.sql` for testing.

File: Classes/Controller/AjaxController.php [hunk 11]
```diff
@@ -405,83 +386,256 @@ public function getConfigurationsAction(ServerRequestInterface $request): Respon
...
+                $rateLimitResult,
+                404,
+            );
+        }
+
+        try {
+            $response = $this->llmServiceManager->chatWithConfiguration($messages, $configuration);
+
+            return $this->jsonResponseWithRateLimitHeaders(
+                CompleteResponse::success($response)->jsonSerialize(),
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply Classes/Controller/AjaxController.php:600]: Fixed:  calls  which applies  server-side for XSS prevention. To restore the HTML for CKEditor, I added a  method in  that uses the standard / pattern to safely reverse the encoding before passing content to CKEditor's data processor and DOMPurify sanitizer.

File: Classes/Domain/DTO/ExecuteTaskRequest.php [hunk 12]
```diff
@@ -0,0 +1,282 @@
...
+     */
+    private const ALLOWED_CONTEXT_SCOPES = [
+        '',
+        'selection',
+        'text',
+        'element',
+        'page',
+        'ancestors_1',
+        'ancestors_2',
+    ];
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply Classes/Domain/DTO/ExecuteTaskRequest.php:64]: This is intentional.  in  defines the content types allowed in the execute-task endpoint (bodytext, header, subheader — i.e. what can be sent as the text to transform).  in  defines what context assembly accepts (tt_content, pages — i.e. record types for surrounding context). They serve different purposes in different request DTOs, so the naming collision is acceptable.
